engineapi/proxy.go

```diff
@@ -0,0 +1,239 @@
+package engineapi
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
+	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+const (
+	ProxyPollInterval = 5 * time.Second
+)
+
+type EngineClientProxy interface {
+	EngineClient
+
+	IsGRPC() bool
+	ResetAndStopProxyClient(*longhorn.InstanceManager) error
+	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
+	Ping() error
+}
+
+// TODO: replace with EngineClientProxy interface later
+type Client interface {
+	IsGRPC() bool
+	ResetAndStopProxyClient(*longhorn.InstanceManager) error
+	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
+	Ping() error
+}
+
+type ProxyHandler struct {
+	Handle map[string]Client
+}
+
+func NewProxyHandler() *ProxyHandler {
+	return &ProxyHandler{
+		Handle: make(map[string]Client),
+	}
+}
+
+func (h *ProxyHandler) GetClient(ownerID string, im *longhorn.InstanceManager, fallBack interface{}) (Client, error) {
```

Comment:
> 1. When the caller tries to get a client, automatically create and add the client to the handler if not exist.

https://github.com/longhorn/longhorn-instance-manager/pull/105#discussion_r867605920

> 2. Modify this/Rename the interface to `CompatibleClient` instead. I mean, automatically **create** the binary client rather than asking the caller to input it if the IM is incompatible. (If the client is the binary client, Longhorn will add arg `--url` during the API invocation rather then during the client init.)
> 3. The main difference between EngineClient and the BackupTargetClient is about the credential. Can we directly use a single interface instead. When callers want to call the backup target related APIs, they need to input the credential.

Not sure if renaming to `CompatibleClient` make sense since the engine binary calls are temporary and [we've talked about deprecate it in the next major version](https://github.com/longhorn/longhorn/pull/3277#discussion_r866467405). Currently we have different inputs to get the binary client (`vc.getEngineClient(volume.Name)`, `GetClientForEngine(engine, m.engines, engine.Status.CurrentImage)`, `getBackupTargetClient(bc.ds, backupTarget)`), the idea is to reuse the current logic to get to  binary client for fallback, so we don't have to worry if the new logic for the binary client going to cause an issue.

Reply:
Updated. Break into `GetCompatibleClient` and `GetClient`.